frontend/javascripts/oxalis/model/sagas/volumetracing_saga.js

```diff
@@ -277,13 +364,140 @@ export function* floodFill(): Saga<void> {
+  resolutionInfo: ResolutionInfo,
+  segmentationCube: DataCube,
+  cellId: number,
+  thirdDimensionOfSlice: number,
+  shouldOverwrite: boolean,
+): void {
+  let currentLabeledVoxelMap: LabeledVoxelsMap = labeledVoxelMapToApply;
+  let thirdDimensionValue = thirdDimensionOfSlice;
+  const get3DAddress = (voxel: Vector2) => {
+    const unorderedVoxelWithThirdDimension = [voxel[0], voxel[1], thirdDimensionValue];
+    const orderedVoxelWithThirdDimension = [
+      unorderedVoxelWithThirdDimension[dimensionIndices[0]],
+      unorderedVoxelWithThirdDimension[dimensionIndices[1]],
+      unorderedVoxelWithThirdDimension[dimensionIndices[2]],
+    ];
+    return orderedVoxelWithThirdDimension;
+  };
+
+  const labeledResolution = resolutionInfo.getResolutionByIndexOrThrow(labeledZoomStep);
+  // Get all available resolutions and divide the list into two parts.
+  const allResolutionsWithIndices = resolutionInfo.getResolutionsWithIndices();
+  // The pivotIndex is the index within allResolutionsWithIndices which refers to
+  // the labeled resolution.
+  const pivotIndex = allResolutionsWithIndices.findIndex(([index]) => index === labeledZoomStep);
+  // `downsampleSequence` contains the current mag and all higher mags (to which
+  // should be downsampled)
+  const downsampleSequence = allResolutionsWithIndices.slice(pivotIndex);
+  // `upsampleSequence` contains the current mag and all lower mags (to which
+  // should be downsampled)
+  const upsampleSequence = allResolutionsWithIndices.slice(0, pivotIndex + 1).reverse();
+
+  // First upsample the voxel map and apply it to all better resolutions.
+  // sourceZoomStep will be higher than targetZoomStep
+  for (const [source, target] of pairwise(upsampleSequence)) {
+    const [sourceZoomStep, sourceResolution] = source;
+    const [targetZoomStep, targetResolution] = target;
+
+    currentLabeledVoxelMap = sampleVoxelMapToResolution(
+      currentLabeledVoxelMap,
+      segmentationCube,
+      sourceResolution,
+      sourceZoomStep,
+      targetResolution,
+      targetZoomStep,
+      dimensionIndices,
+      thirdDimensionOfSlice,
+    );
+    // Adjust thirdDimensionValue so get3DAddress returns the third dimension value
+    // in the target resolution to apply the voxelMap correctly.
+    thirdDimensionValue =
+      Math.floor(thirdDimensionOfSlice / targetResolution[dimensionIndices[2]]) %
+      Constants.BUCKET_WIDTH;
+    const numberOfSlices = Math.ceil(
+      labeledResolution[dimensionIndices[2]] / targetResolution[dimensionIndices[2]],
+    );
+    applyVoxelMap(
+      currentLabeledVoxelMap,
+      segmentationCube,
+      cellId,
+      get3DAddress,
+      numberOfSlices,
+      dimensionIndices[2],
+      shouldOverwrite,
+    );
+  }
+  currentLabeledVoxelMap = labeledVoxelMapToApply;
+
+  // Next we downsamplesample the annotation and apply it.
+  // sourceZoomStep will be lower than targetZoomStep
+  for (const [source, target] of pairwise(downsampleSequence)) {
```

Comment:
It's a little hard to compare but to me it looks like lines 436-469 are almost exactly the same as lines 472-502, except that the `downsampleSequence` is used and the number of slices is set to 1 when calling `applyVoxelMap`. If that's the case I would deduplicate this code.

Reply:
I was a bit hesitant at first to do this, since I was worried that the deduplication could make it more difficult to read, but the result turned out quite nice in my opinion. Thank you for this suggestion! :rocket: